libhb/encavsub.c

```diff
@@ -0,0 +1,408 @@
+        ctx->subtitle->extradata_size = context->extradata_size;
+    }
+
+    if (!realloc_enc_buffer(ctx))
+    {
+        hb_error("encavsubInit: realloc buffer failed");
+        goto fail;
+    }
+    hb_buffer_list_clear(&ctx->list);
+
+    return ctx;
+
+
+fail:
+    if (ctx != NULL)
+    {
+        if (ctx->context != NULL)
+        {
+            avcodec_free_context(&ctx->context);
+        }
+    }
+    free(ctx);
+
+    return NULL;
+}
+
+/***********************************************************************
+ * encavsubClose
+ ***********************************************************************
+ * Close function for libav subtitle encoding that may be wrapped
+ * by HB subtitle encoder
+ **********************************************************************/
+void encavsubClose( hb_encavsub_context_t * ctx )
+{
+    if (ctx == NULL)
+    {
+        return;
+    }
+    avcodec_free_context(&ctx->context);
+    free(ctx);
+}
+
+/***********************************************************************
+ * encavsubWork
+ ***********************************************************************
+ * Work function for libav subtitle encoding that may be wrapped
+ * by HB subtitle encoder
+ **********************************************************************/
+int encavsubWork( hb_encavsub_context_t * ctx,
+                  hb_buffer_t ** buf_in,
+                  hb_buffer_t ** buf_out )
+{
+    hb_buffer_t * in = *buf_in;
+    AVSubtitle    subtitle;
+    int           num_rects = 1;
+    int64_t       duration;
+
+    if (in->s.flags & HB_BUF_FLAG_EOF)
+    {
+        /* EOF on input stream - send it downstream & say that we're done */
+        *buf_in = NULL;
+
+        *buf_out = in;
+        return HB_WORK_DONE;
+    }
+
+    // Create an AVSubtitle from the input buffer
+    memset(&subtitle, 0, sizeof(subtitle));
+    subtitle.rects = av_mallocz_array(num_rects, sizeof(subtitle.rects));
+    if (subtitle.rects == NULL)
+    {
+        *buf_out = hb_buffer_eof_init();
+        hb_error("encavsubInit: av_mallocz_array failed");
+        return HB_WORK_DONE;
+    }
+
+    int ii;
+    for (ii = 0; ii < num_rects; ii++)
+    {
+        subtitle.rects[ii]  = av_mallocz(sizeof(*subtitle.rects[0]));
+        if (!subtitle.rects[ii]) {
+            avsubtitle_free(&subtitle);
+            *buf_out = hb_buffer_eof_init();
+            hb_error("encavsubInit: av_mallocz failed");
+            return HB_WORK_DONE;
+        }
+        subtitle.num_rects++;
+    }
+
+    if (ctx->subtitle->format == TEXTSUB)
+    {
+        subtitle.rects[0]->type = SUBTITLE_ASS;
+        subtitle.rects[0]->ass  = av_strdup((char*)in->data);
+        if (subtitle.rects[0]->ass == NULL)
+        {
+            avsubtitle_free(&subtitle);
+            *buf_out = hb_buffer_eof_init();
+            hb_error("encavsubInit: av_strdup failed");
+            return HB_WORK_DONE;
+        }
+    }
+    else
+    {
+        // TODO: bitmap subtitles
```

Comment:
TODO :)

Reply:
encavsub is currently only used for conversion to tx3g.  My plan was to flesh out the bitmap path by writing a pgs encoder for ffmpeg and using it to convert vobsub to pgs.  The pgs encoder is  written for the most part, but there's work left to do (like getting it through ffmpeg review).